keymanapp/keyman
# fix(windows): ensure all modifier events go to serialized queue

Looking at the logs it seems to indicate it occurs when switching between keyman and non-keyman keyboards. The use of <kbd>Alt</kbd> + <kbd>Tab</kbd> to switch applications. #6392 made a change that meant key events were only sent to the keyman serialized queue when a Keyman keyboard was active. This resolved the issue with EMOJI input pop-up. However, it is likely the cause as to why the modifier keys are sometimes being seen out of sequence.

This change makes it so that all modifier events are sent to the serialized server so that the modifier state does not lose synchronization.

# USER TESTING :

This PR changes code around modifier keys. Therefore the tests need to cover interactions around these keys being pressed.
Unfortunately, I was not able to find a way to reproduce the behaviour to test it is resolved. We need to make sure though that
there are no regressions.



**GROUP_WIN10**  - Run the tests in a Windows 10 OS
**GROUP_WIN11** - Run the tests in Windows 11 OS


**TEST_ALTGR_IN_RULES** 
This is to test the key rules which contain <kbd>AltGr</kbd> or <kbd>Shift</kbd> + <kbd>AltGr</kbd> still work correctly.

* Load a keyboard `khmer_angkor` 
* Open LibreOffice or Notepad 
* Type a <kbd>AltGr</kbd> + <kbd>o</kbd> 
* Observe output of  ឱ​

* Type a <kbd>Shift</kbd> + <kbd>AltGr</kbd> + <kbd>o</kbd> 
* Observe output of ᧨


**TEST_HOTKEYS**
This is to test switching between non-keyman and keyman keyboards using hotkeys. To make sure this PR has not broken Hotkeys that use modifiers 
* Install two different keyman keyboards in addition to the default for the Windows installation.
* Use the Keyman Configuration to assign hotkeys to the default Keyboard as well as the two Keyman Keyboards that use modifiers for the hotkey for example <kbd>Ctrl</kbd> + <kbd>1</kbd>  or <kbd>Alt</kbd> + <kbd>K</kbd>
* Open Notepad 
* Use the hotkeys to switch between keyboards
* Verify that switching between the keyboards is working


**TEST_HOTKEYS_SWITCH_APPS** 
This extends the previous test to test switching between non-keyman and keyman keyboards using hotkeys and also using <kbd>Alt</kbd> + <kbd>Tab</kbd> to switch applications.
* Install two different keyman keyboards in addition to the default for the Windows installation.
* Use the Keyman Configuration to assign hotkeys to the default Keyboard as well as the two Keyman Keyboards
* Open two text input applications for example Libre Office and Notepad.
* Turn on the OSK keyboard
* Open Libre Office and select the default Keyboard with the Hotkeys type some text.
* Use <kbd>Alt</kbd> + <kbd>Tab</kbd> to switch to NotePad
* Use the Hotkeys to switch to a Keyman Keyboard and observe you can use that keyboard.
* Observe that on the <kbd>Alt</kbd> is not pressed down "stuck" on.
* Use the HotKeys to switch to the second Keyman keyboard
* Use <kbd>Alt</kbd> + <kbd>Tab</kbd> to switch to LibreOffice
* Type and observe the keyboard works correctly 
* Observe that the <kbd>Alt</kbd>does not get stuck
* Try a combination of switching between keyboards and applications. Pressing different modifier keys etc.
* We want to make use the ALT Key or any other modifier key does not stay down - become stuck.

**TEST_WIN_EMOJI_INPUT** 

* Start Keyman
* Select a non-keyman keyboard
* Open Notepad 
* Type some text on the physical keyboard. 
* Press the <kbd>&#x229E; Win</kbd> + <kbd>.</kbd> to bring up the emoji and gif window.
* In the search box type some characters.
* Does it allow input into the emoji panels text field? Verify that input still works and is not completely blocked.
